streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1211,52 +1224,95 @@ object ZSink extends ZSinkPlatformSpecific {
+    max: Long
+  )(f: (S, A) => ZIO[R1, E1, S]): ZSink[R1, E1, A, A, S] =
+    foldWeightedDecomposeM[R, R1, E1, E1, A, S](z)(costFn, max, (a: A) => UIO.succeed(Chunk.single(a)))(f)
+
+  /**
+   * Creates a sink that effectfully folds elements of type `A` into a structure
+   * of type `S`, until `max` worth of elements (determined by the `costFn`) have
+   * been folded.
+   */
+  final def foldWeightedDecomposeM[R, R1 <: R, E, E1 >: E, A, S](
+    z: S
+  )(
+    costFn: A => ZIO[R, E, Long],
+    max: Long,
+    decompose: A => ZIO[R, E, Chunk[A]]
+  )(f: (S, A) => ZIO[R1, E1, S]): ZSink[R1, E1, A, A, S] =
     new ZSink[R1, E1, A, A, S] {
-      type State = (S, Long, Boolean, Chunk[A])
+      import internal.FoldWeightedState
 
-      val initial = UIO.succeed((z, 0L, true, Chunk.empty))
+      type State = FoldWeightedState[S, A]
+
+      val initial = UIO.succeed(FoldWeightedState(z, 0L, true, Chunk.empty))
 
       def step(state: State, a: A) =
         costFn(a).flatMap { cost =>
-          val newCost = cost + state._2
+          val newCost = cost + state.cost
 
-          if (newCost > max) UIO.succeed((state._1, state._2, false, Chunk.single(a)))
-          else if (newCost == max) {
-            f(state._1, a).map((_, newCost, false, Chunk.empty))
-          } else f(state._1, a).map((_, newCost, true, Chunk.empty))
+          if (newCost > max)
+            decompose(a).map(leftovers => state.copy(cont = false, leftovers = state.leftovers ++ leftovers))
+          else if (newCost == max)
+            f(state.s, a).map(FoldWeightedState(_, newCost, false, Chunk.empty))
+          else
+            f(state.s, a).map(FoldWeightedState(_, newCost, true, Chunk.empty))
         }
 
-      def extract(state: State) = UIO.succeed((state._1, state._4))
+      def extract(state: State) = UIO.succeed((state.s, state.leftovers))
 
-      def cont(state: State) = state._3
+      def cont(state: State) = state.cont
     }
 
   /**
    * Creates a sink that folds elements of type `A` into a structure
    * of type `S`, until `max` worth of elements (determined by the `costFn`)
    * have been folded.
+   *
+   * @note In case
    */
   final def foldWeighted[A, S](
     z: S
-  )(costFn: A => Long, max: Long)(f: (S, A) => S): ZSink[Any, Nothing, A, A, S] =
+  )(costFn: A => Long, max: Long)(
+    f: (S, A) => S
+  ): ZSink[Any, Nothing, A, A, S] =
+    foldWeightedDecompose(z)(costFn, max, Chunk.single[A](_))(f)
+
+  /**
+   * Creates a sink that folds elements of type `A` into a structure
+   * of type `S`, until `max` worth of elements (determined by the `costFn`)
+   * have been folded.
+   *
+   * @note In case
+   */
+  final def foldWeightedDecompose[A, S](
+    z: S
+  )(costFn: A => Long, max: Long, decompose: A => Chunk[A])(
+    f: (S, A) => S
+  ): ZSink[Any, Nothing, A, A, S] =
     new SinkPure[Nothing, A, A, S] {
-      type State = (S, Long, Boolean, Chunk[A])
+      import internal.FoldWeightedState
```

Comment:
Since we are starting to use `ZSink.internal` for complex states, it makes sense to restrict the old imports to only the needed types. Just a mental note.